ui/app/hooks/useTokenTracker.js

```diff
@@ -0,0 +1,95 @@
+import { useState, useEffect, useRef, useCallback } from 'react'
+import TokenTracker from '@metamask/eth-token-tracker'
+import { isEqual } from 'lodash'
+import { useSelector } from 'react-redux'
+import { getCurrentNetwork, getSelectedAddress } from '../selectors'
+import { getTokens } from '../ducks/metamask/metamask'
+
+
+export function useTokenTracker () {
+  const network = useSelector(getCurrentNetwork)
+  const userAddress = useSelector(getSelectedAddress)
+  // use `isEqual` comparison function because the token array is serialized
+  // from the background so it has a new reference with each background update,
+  // even if the tokens haven't changed
+  const tokens = useSelector(getTokens, isEqual)
+
+  const [loading, setLoading] = useState(() => tokens?.length >= 0)
+  const [tokensWithBalances, setTokensWithBalances] = useState([])
+  const [error, setError] = useState(null)
+  const tokenTracker = useRef(null)
+
+  const updateBalances = useCallback((tokensWithBalances) => {
+    setLoading(false)
+    setTokensWithBalances(tokensWithBalances)
```

Comment:
Nit: I think it would make more sense for this to _follow_ `setTokensWithBalances`, so that there's no danger of the list being rendered before the balances are set.

These two state updates are _probably_ always batched together in practice? 🤔 But I'm not sure, I don't think there's any guarantee that is the case.

```suggestion
    setTokensWithBalances(tokensWithBalances)
    setLoading(false)
```

Reply:
... no call order guarantee? That doesn't seem right 🤔 I can't imagine this stuff would happen out-of-order - it'd be impossible to ensure the state was valid if that was the case.

Reply:
Huh, so you're definitely correct.

I thought at first I was confusing useState with useEffect which I thought had its guarantees of order removed to support suspense/async rendering.

Turns out in both cases call order is important and guaranteed. In this case the updates are batched together and only a single render happens -- but if you were to setState inside a promise like this each setState would trigger a new render, in order of the setState calls.

Technical details aside it just makes sense to put them in the order you outlined